pkg/globalnet/controllers/cluster_egressip_controller.go

```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+
+			return true
+		}
+
 		return false
 	}
 
-	return allocateIPs(key, egressIP.Spec.NumberOfIPs, c.pool, &egressIP.Status)
+	return c.allocateGlobalIPs(key, numberOfIPs, status)
 }
 
-func (c *clusterGlobalEgressIPController) onRemove(egressIP *submarinerv1.ClusterGlobalEgressIP) bool { // nolint unparam
-	// TODO - remove IP table rules for the allocated IPs
+func (c *clusterGlobalEgressIPController) onUpdate(key string, numberOfIPs *int, status *submarinerv1.GlobalEgressIPStatus) bool {
+	if *numberOfIPs == len(status.AllocatedIPs) {
+		klog.V(log.DEBUG).Infof("Update called for %q, but numberOfIPs %q are already allocated", key, *numberOfIPs)
+		return false
+	}
+
+	// If numGlobalIPs is modified, delete the existing allocation.
+	if len(status.AllocatedIPs) > 0 {
+		c.flushClusterGlobalEgressRules(status)
+		_ = c.pool.Release(status.AllocatedIPs...)
+	}
+
+	return c.allocateGlobalIPs(key, numberOfIPs, status)
+}
+
+func (c *clusterGlobalEgressIPController) onRemove(key string, egressIP *submarinerv1.ClusterGlobalEgressIP) bool { // nolint unparam
+	if len(egressIP.Status.AllocatedIPs) > 0 {
+		c.flushClusterGlobalEgressRules(&egressIP.Status)
+		_ = c.pool.Release(egressIP.Status.AllocatedIPs...)
+	}
+
+	return false
+}
+
+func (c *clusterGlobalEgressIPController) flushClusterGlobalEgressRules(status *submarinerv1.GlobalEgressIPStatus) {
+	snatIP := c.getTargetSNATIPaddress(status.AllocatedIPs)
+	if snatIP == "" {
+		klog.Warning("flushClusterGlobalEgressRules called with 0 AllocatedIPs")
+		return
+	}
+
+	c.deleteClusterGlobalEgressRules(c.localSubnets.Elements(), snatIP)
+}
+
+func (c *clusterGlobalEgressIPController) deleteClusterGlobalEgressRules(srcIPList []string, snatIP string) {
+	for _, srcIP := range srcIPList {
+		if err := c.iptIface.RemoveClusterEgressRules(srcIP, snatIP, globalNetIPTableMark); err != nil {
+			klog.Errorf("Error while cleaning up ClusterEgressIPs: %v", err)
+		}
+	}
+}
+
+func (c *clusterGlobalEgressIPController) programClusterGlobalEgressRules(status *submarinerv1.GlobalEgressIPStatus) error {
+	snatIP := c.getTargetSNATIPaddress(status.AllocatedIPs)
+	if snatIP == "" {
+		klog.Warning("programClusterGlobalEgressRules called with 0 AllocatedIPs")
+		return nil
+	}
+
+	egressRulesProgrammed := []string{}
+
+	for _, srcIP := range c.localSubnets.Elements() {
+		if err := c.iptIface.AddClusterEgressRules(srcIP, snatIP, globalNetIPTableMark); err != nil {
+			c.deleteClusterGlobalEgressRules(egressRulesProgrammed, snatIP)
+
+			return err
+		}
+
+		egressRulesProgrammed = append(egressRulesProgrammed, srcIP)
+	}
+
+	return nil
+}
+
+func (c *clusterGlobalEgressIPController) getTargetSNATIPaddress(allocIPs []string) string {
+	var snatIP string
+
+	allocatedIPs := len(allocIPs)
+
+	if allocatedIPs == 1 {
+		snatIP = allocIPs[0]
+	} else if allocatedIPs > 1 {
+		snatIP = fmt.Sprintf("%s-%s", allocIPs[0], allocIPs[len(allocIPs)-1])
+	}
```

Comment:
Again, let's assume > 0 to avoid an unused code path.

```suggestion
	} else {
		snatIP = fmt.Sprintf("%s-%s", allocIPs[0], allocIPs[len(allocIPs)-1])
	}
```